# Exploring eBay Car Sales Data

## Introduction and Purpose 

We will be working with a dataset of used cars from the classifieds section of the German eBay website.

The aim of which is to clean the data and analyze the included used car listings. 

The dataset can be found [here](https://data.world/data-society/used-cars-data).  

Along with the dataset, the we will use the provided column dictionary:
- **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
- **name** - Name of the car.
- **seller** - Whether the seller is private or a dealer.
- **offerType** - The type of listing
- **price** - The price on the ad to sell the car.
- **abtest** - Whether the listing is included in an A/B test.
- **vehicleType** - The vehicle Type.
- **yearOfRegistration** - The year in which the car was first registered.
- **gearbox** - The transmission type.
- **powerPS** - The power of the car in PS.
- **model** - The car model name.
- **kilometer** - How many kilometers the car has driven.
- **monthOfRegistration** - The month in which the car was first registered.
- **fuelType** - What type of fuel the car uses.
- **brand** - The brand of the car.
- **notRepairedDamage** - If the car has a damage which is not yet repaired.
- **dateCreated** - The date on which the eBay listing was created.
- **nrOfPictures** - The number of pictures in the ad.
- **postalCode** - The postal code for the location of the vehicle.
- **lastSeenOnline** - When the crawler saw this ad last online.





Let's start by importing the libraries and reading the dataset into pandas.
## Reading in and Exploring the Dataset

In [1]:
import pandas as pd 
import numpy as ny

#Now we will read our csv file into pandas
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

Next, let's explore the dataset.

In [3]:
print(autos.info())
print('\n')
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In examining this dataset, it looks like we will be working with a 50,000 entries with 20 descriptive columns - a mixture of numerical and string datatypes. Also, it looks like some columns have some non-null values, most likely blank or unfilled data due to scraping this data from online.

Also, it looks like some column names are fully joined into one word vs. spaced with underscores. 

Let's work to convert the column name to this way along with rewording based on the column descriptions from above.

## Cleaning the Data

Examining the columns, let's convert all the names from camel type to snake type where applicable. 

I will also rename some columns for clarity:
- rename 'gearbox' to 'transmission_type' for clarity 
- rename 'powerPS' to 'horse_power' as PS is an abbreviation for the German word for 'horsepower'
- shortening the woring for 'notRepairedDamage' and 'nrOfPictures'

In [4]:
#Let's use the columns attribute to assign our column names
autos_cols = autos.columns
print(autos_cols)

#rename the columns
autos_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'transmission_type', 'horse_power', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'postal_code',
       'last_seen']

#now reassign the new columns back to the dataframe
autos.columns = autos_cols

#check
print(autos.head())

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1

Let's continue to do some exploration and cleaning.

## Continuing to Explore and Clean the Data

As we look to clean the data, we will look for:
- Text columns where all or almost all values are the same. These can be dropped for our analysis.
- Examples of numeric data stored as text which we can clean and convert.

In [5]:
#use df.describe method to explore the stats of our columns
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 19:48:02  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year transmission_type   horse_power  model  \
count         44905       50000.000000             47320  50000.000000  47242   
unique            8                NaN                 2           NaN    245   
top       limousine                NaN           manuell           NaN   golf   
freq          12859                NaN             36993           NaN   4024   
mean            NaN        2005.073280               NaN    116.355920    NaN   
std             NaN         105.712813               NaN    209.216627    NaN   
min             NaN        1000.000000               NaN      0.000000    NaN   
25%             NaN        1999.000000               NaN     70.000000    NaN   
50%             NaN        2003.000000               NaN    105.000000    NaN   
75%             NaN        2008.000000               NaN    150.000000    NaN   
max             NaN        9999.000000               NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Based on the descriptive data, it looks like the seller, offer_type, ab_test, transmission_type and unrepaired_damage seem to have mostly one value as they only have 2 unique instances. 

Also, some of these columns that would be expected to have numerical data show NaN, and are probably written as strings such as price, odemeter, horse_power, num_of_pictures, registration_year and registration_month. We will need to explore and most likely clean these columns to conver to numerical data.

In [6]:
#let's look at what kind of data some of these columns hold
print(autos['seller'].value_counts())
print(autos['seller'].unique())
print('\n')
print(autos['offer_type'].value_counts())
print(autos['offer_type'].unique())
print('\n')
print(autos['ab_test'].value_counts())
print(autos['ab_test'].unique())
print('\n')
print(autos['transmission_type'].value_counts())
print(autos['transmission_type'].unique())
print('\n')
print(autos['unrepaired_damage'].value_counts())
print(autos['unrepaired_damage'].unique())
print('\n')
print(autos['num_of_pictures'].value_counts())
print(autos['num_of_pictures'].unique())


privat        49999
gewerblich        1
Name: seller, dtype: int64
['privat' 'gewerblich']


Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
['Angebot' 'Gesuch']


test       25756
control    24244
Name: ab_test, dtype: int64
['control' 'test']


manuell      36993
automatik    10327
Name: transmission_type, dtype: int64
['manuell' 'automatik' nan]


nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64
['nein' nan 'ja']


0    50000
Name: num_of_pictures, dtype: int64
[0]


After doing some initial exploration of some of the columns that seem to have one or two types of data, we can look to remove these from our analysis for the time being.

In [7]:
#Because num_of_pictures column only has '0' in the column, let's remove
autos = autos.drop('num_of_pictures',axis=1)
print('Columns: ', autos.columns) #test

Columns:  Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'transmission_type', 'horse_power',
       'model', 'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


In [8]:
#Because seller also has one type of data, we will drop that as well
autos = autos.drop('seller', axis=1)
print('Columns: ', autos.columns)#test2

Columns:  Index(['date_crawled', 'name', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'transmission_type', 'horse_power',
       'model', 'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


In [9]:
#Finally, because offer_typer only has one type of value, let's remove that
autos = autos.drop('offer_type', axis=1)
print('Columns: ', autos.columns)#test3

Columns:  Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'transmission_type', 'horse_power', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


Now, let's put our attention to the price and odemeter columns as they contain numeric data stored as strings. For each of these, we will:
- Remove any non-numeric data
- Convert the column to numeric dtype
- Remove the column to odometer_km

In [10]:
#first let's explore the non-numeric data in price
autos['price'].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [11]:
#Looks like it follows the patter: $ letter, number, comma, numbers
#let's begin to remove these non-numerical characters 
autos['price'] = autos['price'].str.replace('$','').str.replace(',','')
autos['price']= autos['price'].astype(int)
autos['price'].value_counts()#check

0         1421
500        781
1500       734
2500       643
1000       639
1200       639
600        531
800        498
3500       498
2000       460
999        434
750        433
900        420
650        419
850        410
700        395
4500       394
300        384
2200       382
950        379
1100       376
1300       371
3000       365
550        356
1800       355
5500       340
1250       335
350        335
1600       327
1999       322
          ... 
46200        1
29600        1
13480        1
21700        1
7373         1
3279         1
4286         1
188          1
17830        1
9130         1
910          1
238          1
2671         1
69900        1
151990       1
2479         1
4510         1
86500        1
47499        1
16998        1
27299        1
41850        1
4780         1
686          1
6495         1
20790        1
8970         1
846          1
2895         1
33980        1
Name: price, Length: 2357, dtype: int64

Great, looks like it worked, now let's do the same for the odometer column

In [12]:
#explore
autos['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [13]:
#the pattern seems to follow: numerical, comma, numerical, km
autos['odometer'] = autos['odometer'].str.replace(',','').str.replace('km','')
autos['odometer']= autos['odometer'].astype(int)
autos['odometer'].unique() #check
autos['odometer'].value_counts()#check

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

In [14]:
#now let's rename the odometer column to reflect its in kilometers
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)
autos.columns #check

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'transmission_type', 'horse_power', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

Now that we have taken out columns where almost all the values are the same (seller, offer_type, num_of_pictures) and converted price and odometer_km to numeric types, let's continue to explore these columns by 
analyzing the columns using min and max values and look for any values that look like outliers we should remove.

In [15]:
#First look at odomter_km column
print(autos['odometer_km'].unique().shape) 
print('\n')
print(autos['odometer_km'].describe())
print('\n')
print(autos['odometer_km'].value_counts().head())
print('\n')
print(autos['odometer_km'].value_counts().sort_index())

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64


5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64


In [16]:
#Now let's look at the price column
print(autos['price'].unique().shape) 
print('\n')
print(autos['price'].describe())
print('\n')
print(autos['price'].value_counts().sort_index())
print('\n')
print(autos['price'].value_counts().sort_index(ascending=False))

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
26

Looking at the price column, we can see that there are some cars that probably are given for free (that's possible). But there are also some suspiciously expensive car at 99999999 and another one at  . Those two are most likely outliers.Lets use the 1.5xIQR statistics rule to see if more outliers exist 

In [17]:
IQR = 7.200000e+03 - 1.100000e+03

In [18]:
low_outliers = 1.100000e+03 - (1.5*IQR)
high_outliers = 7.200000e+03 + (1.5*IQR)

In [19]:
print(low_outliers, high_outliers)

-8050.0 16350.0


Based on this and the wide spread of prices, we probably shouldn't use this to determine outliers. Those looking at the data again, it looks like the prices climb pretty uniformly until we get to $350000, then it jumps to $999,990. We will make the cutoff $350000 and make the rest above that outliers 

In [20]:
autos[autos["price"] > 350000] = ny.NaN

In [21]:
autos['price'].value_counts().sort_index(ascending=False).head(10)

350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
250000.0    1
220000.0    1
198000.0    1
197000.0    1
Name: price, dtype: int64

In [22]:
#remove the rows
autos.dropna()

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
30     2016-03-14 11:47:31                         Peugeot_206_Unfallfahrzeug   
31     2016-03-14 16:53:09                          Noch_gut_erhaltenen_C_320   
33     2016-03-15 12:25:26  Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...   
35     2016-03-29 21:52:56                                        Ford_Mondeo   
36     2016-03-16 23:46:31          Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
38     2016-03-21 15:51:10                        Volkswagen_Golf_1.4_Special   
39     2016-03-11 13:58:38                 Mercedes_C_180_Kombi__zu_verkaufen   
40     2016-03-07 14:50:03  VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...   
43     2016-03-08 16:51:22           Audi_A1_1.2_TFSI_Attraction_mit_Garantie   
44     2016-03-12 10:38:06                                       Zu_verkaufen   
45     2016-03-12 17:39:55                                 Toyota_Starlet_1.3   
...                    ...                                                ...   
49959  2016-03-28 21:53:21           Mercedes_Benz_C_180_T_Kompressor_Classic   
49962  2016-03-14 17:57:15  Mitsubishi_Space_Star_1__3_L__Bj_2004_Standhei...   
49963  2016-03-26 19:52:59            Mercedes_Benz_B_200_CDI_Special_Edition   
49965  2016-03-11 10:53:51          Opel_Astra_1.6_Lenkradheizung~Sitzheizung   
49966  2016-04-02 19:49:19  Citroën_C1_1.0_**Euro4**TÜV_OKT_2017**Scheiten...   
49968  2016-04-01 17:49:15                  Mercedes_Benz_190_D_2.5_Automatik   
49969  2016-03-17 18:49:02             Nissan_X_Trail_2.2_dCi_4x4_Sport_m.AHZ   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49981  2016-03-15 09:38:21             Opel_A

Looks like we have now reduced our rows to 35,043 from 50,0000. 
Now let's move onto the date columns and the date range it covers.

There are five columns that holds this data, differentiated by what created them: a web crawler or the website itself.

Right now, the date_crawled, last_seen and ad_created are string values. We need to convert them to numerical values. The other two, registration_month and registration_year, are numeric values.

Let's explore the three string columns formatting. 

In [23]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

It looks to be in the %Y-%m-%d %H:%M:%S format

Let's explore the date range, by selecting the first 10 characters in each column.

In [24]:
#let's extract the date values, generate a distro, then sort the index
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print('\n')
print("Min Values ", autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).tail())
print('\n')
print('Max Values ', autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).head())

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03326
2016-03-09    0.03320
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03676
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03150
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03748
2016-03-22    0.03290
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03414
2016-03-30    0.03362
2016-03-31    0.03190
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03648
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
NaN           0.00028
Name: date_crawled, dtype: float64


Min Values  2016-04-05    0.01310
2016-03-18    0.01306
2016-04-06    0.00318
2016-04-07    0.00142
NaN           0.00028
Name: date_crawled, dtype: float64


Max Values  2016-04-03    0.03868


From our date_crawled column, we can conclude that:
- 04-07-2016 has the lowest distribution at 0.142%
- 03-30-2016 has the highest distribution at 3.782%
- NaN has a distribution of 0.028%

In [25]:
#Now let's look at the ad_created column
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print('\n')
print("Min Values ", autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).tail())
print('\n')
print('Max Values ', autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).head())

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
2015-12-30    0.00002
2016-01-03    0.00002
2016-01-07    0.00002
2016-01-10    0.00004
2016-01-13    0.00002
2016-01-14    0.00002
2016-01-16    0.00002
2016-01-22    0.00002
2016-01-27    0.00006
2016-01-29    0.00002
2016-02-01    0.00002
2016-02-02    0.00004
2016-02-05    0.00004
2016-02-07    0.00002
2016-02-08    0.00002
2016-02-09    0.00004
2016-02-11    0.00002
2016-02-12    0.00006
2016-02-14    0.00004
2016-02-16    0.00002
2016-02-17    0.00002
2016-02-18    0.00004
2016-02-19    0.00006
2016-02-20    0.00004
2016-02-21    0.00006
               ...   
2016-03-10    0.03186
2016-03-11    0.03278
2016-03-12    0.03660
2016-03-13    0.01692
2016-03-14    0.03522
2016-03-15    0.03374
2016-03-16    0.03000
2016-03-17    0.03118
2016-03-18    0.01372
2016-03-19    0.03384
2016-03-20    0.03786
2016-03-21    0.03768
2016-03-22    0.03276
2016-03-23    0.03218
2016-03-24

From our ad_created column, we can conclude that:

- 02-01-2016, 01-14-2016, 01-07-2016, and 02-22-2016 have the lowest distribution at 0.002%
- 03-20-2016 has the highest distribution at 3.786%

In [26]:
#Finally, let's look at our last_seen column
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print('\n')
print("Min Values ", autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).tail())
print('\n')
print('Max Values ', autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).head())

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00758
2016-03-09    0.00984
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02380
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02072
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02084
2016-03-29    0.02232
2016-03-30    0.02484
2016-03-31    0.02382
2016-04-01    0.02310
2016-04-02    0.02488
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12424
2016-04-06    0.22092
2016-04-07    0.13092
NaN           0.00028
Name: last_seen, dtype: float64


Min Values  2016-03-18    0.00742
2016-03-07    0.00536
2016-03-06    0.00442
2016-03-05    0.00108
NaN           0.00028
Name: last_seen, dtype: float64


Max Values  2016-04-06    0.22092
2016-0

From our last_seen column, we can conclude that:

- 03-05-2016 has the lowest distribution at 0.108%
- 04-07-2016 has the highest distribution at 13.092%
- Nan has a distribution of 0.028%

Now let's understand the distribution of registration_year

In [27]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Knowing these are years, we know some of this data is invalid as the min year is 1000, max is 9999, with a standard deviation of 105 - quite the spread of years. We need to clean this column.

Because a car can't be registered after a listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest is tougher. Realistically, it could be within the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 internval and see if it's safe to remove those rows.

In [28]:
autos['registration_year'].value_counts().sort_index()

1000.0       1
1001.0       1
1111.0       1
1500.0       1
1800.0       2
1910.0       9
1927.0       1
1929.0       1
1931.0       1
1934.0       2
1937.0       4
1938.0       1
1939.0       1
1941.0       2
1943.0       1
1948.0       1
1950.0       3
1951.0       2
1952.0       1
1953.0       1
1954.0       2
1955.0       2
1956.0       5
1957.0       2
1958.0       4
1959.0       7
1960.0      33
1961.0       6
1962.0       4
1963.0       9
          ... 
2001.0    2702
2002.0    2533
2003.0    2727
2004.0    2737
2005.0    3015
2006.0    2707
2007.0    2304
2008.0    2231
2009.0    2097
2010.0    1597
2011.0    1634
2012.0    1323
2013.0     806
2014.0     665
2015.0     399
2016.0    1316
2017.0    1452
2018.0     491
2019.0       3
2800.0       1
4100.0       1
4500.0       1
4800.0       1
5000.0       4
5911.0       1
6200.0       1
8888.0       1
9000.0       2
9996.0       1
9999.0       4
Name: registration_year, Length: 97, dtype: int64

Using the 2016 as our high cutoff year, let's look at the above sorted counts. The lowest realistic year is 1910 before it jumps to 1800. Let's use 1910 as our lower parameter. 

In [29]:
autos = autos[autos['registration_year'].between(1910,2016)]

In [30]:
print(autos['registration_year'].value_counts(normalize=True))
print('\n')
print(autos['registration_year'].value_counts(normalize=True).head())
print('\n')
print(autos['registration_year'].value_counts(normalize=True).tail())

2000.0    0.069852
2005.0    0.062792
1999.0    0.062438
2004.0    0.057002
2003.0    0.056794
2006.0    0.056377
2001.0    0.056273
2002.0    0.052753
1998.0    0.051087
2007.0    0.047984
2008.0    0.046464
2009.0    0.043673
1997.0    0.042236
2011.0    0.034030
2010.0    0.033260
1996.0    0.030073
2012.0    0.027553
2016.0    0.027408
1995.0    0.027324
2013.0    0.016786
2014.0    0.013850
1994.0    0.013745
1993.0    0.009268
2015.0    0.008310
1990.0    0.008226
1992.0    0.008122
1991.0    0.007414
1989.0    0.003770
1988.0    0.002957
1985.0    0.002166
            ...   
1977.0    0.000458
1966.0    0.000458
1975.0    0.000396
1969.0    0.000396
1965.0    0.000354
1964.0    0.000250
1910.0    0.000187
1963.0    0.000187
1959.0    0.000146
1961.0    0.000125
1956.0    0.000104
1962.0    0.000083
1958.0    0.000083
1937.0    0.000083
1950.0    0.000062
1934.0    0.000042
1941.0    0.000042
1951.0    0.000042
1954.0    0.000042
1955.0    0.000042
1957.0    0.000042
1952.0    0.

Looking at the new distribution after dropping years that were above 2016 and below 1920, we conclude the following:
- 2000 has the highest distribution at 6.9852%: 1998 - 2009 seens the most offers. 
- Years between 1927 and 1939 have the smallest at 0.0021%

Now let's explore Price by brand. We will use aggreagation on this column by:
- ID the unique values we want to aggregate by
- Create an empty dictionary to store the aggregate data
- Calculate the mean of whichever column we are interested in
- Assign the val/mean to the dict as k/v

In [31]:
#explore unique values of the brand column
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [32]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
citroen           0.013912
smart             0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

Based on this data, it looks like the top 6 car brands also hold the highest distribution, above 5%. 
 - Volkswagen is the highest at 21%
 - BMW is at 11%
 - Opel at 11%
 - Mercedez Benz is at 9.5%
 - Audi is at 8%
 - Ford is at 6% 
 
 Not a total surprise that the German made cars have the highest distribution on German eBay. We will aggregate by the above. 
 

In [33]:
brand_index = autos['brand'].value_counts(normalize=True).index[:6]
print(brand_index)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [34]:
mean_price_for_top_6 = {}
for b in brand_index:
    top_brands = autos[autos['brand']==b]
    mean_price = top_brands['price'].mean()
    mean_price_for_top_6[b] = mean_price
    
for b in mean_price_for_top_6:
    print(b, ':', mean_price_for_top_6[b])

audi : 9093.65003615329
volkswagen : 5231.081983308787
opel : 2876.716403542549
ford : 3652.095223880597
bmw : 8102.536248343744
mercedes_benz : 8485.239571958942


Looking at the above, out of the most posted car brands, Audi is on average, the most expensive at 9093.65, followed by Mercedes at 8485.23. The cheapest car of the bunch is Opel at 2876. 

The German cars are both the highest quantity posted and the more expensive of the cohort. 

|Brand|Mean Price|
|-----|----------|
|Audi | 9093.65|
|Mercedes| 8485.24 |
|BMW  | 8102.54 |
|Volkswagen | 5231.08 |
|Ford| 3652.09 |
|Opel | 2876.71 |


Now that we have aggregated across brands to understand mean price, we now understand that Audi, BMW and Mercedez Benz are more expensive, Ford and Opel are less expensive and Volkswagen in between. 

Let's use aggregation to now understand the average mileage for those cars if there is a link to mean price. We will combine the data from both series objects into a single dataframe (with a shared index) and display it directly. 

In [35]:
#mean price and mileage for top 6 brands
mean_mileage_for_top_6 = {}
mean_price_for_top_6 = {}
for b in brand_index:
    top_brands = autos[autos['brand']==b]
    mean_price = top_brands['price'].mean()
    mean_mileage = top_brands['odometer_km'].mean()
    mean_price_for_top_6[b] = mean_price
    mean_mileage_for_top_6[b] = mean_mileage


#convert dictionaries to series
bmp = pd.Series(mean_price_for_top_6)
bmm = pd.Series(mean_mileage_for_top_6)

#create a dataframe from first series 
brand_df = pd.DataFrame(bmp, columns=['mean_price'])

#adding a columnd to the dataframe
brand_df['mean_mileage'] = bmm
brand_df

brand_df.sort_values(by='mean_price', ascending=False)


mean_price   mean_mileage
audi           9093.650036  129287.780188
mercedes_benz  8485.239572  130856.082114
bmw            8102.536248  132431.383684
volkswagen     5231.081983  128724.104075
ford           3652.095224  124068.656716
opel           2876.716404  129223.142087

The top three brands in terms of price, on average, have a higher average mileage. Though as a trend, the correlation between price and mileage seems weak and thus, the main driver of price between the two seem to relate more closely to the brand of the car itself. It seems that Audi, Mercedes and BMW are seen by people as the most premium of the car brands as their price is higher relative to the mileage being around the same. 

## Further Data Cleaning and Analysis

In [36]:
autos.head(25)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20  2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21  2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22  2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23  2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24  2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25  2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   

      price  ab_test vehicle_type  registration_year transmission_type  \
0    5000.0  control          bus             2004.0           manuell   
1    8500.0  control    limousine             1997.0         automatik   
2    8990.0     test    limousine             2009.0           manuell   
3    4350.0  control   kleinwagen             2007.0         automatik   
4    1350.0     test        kombi             2003.0           manuell   
5    7900.0     test          bus             2006.0         automatik   
6     300.0     test    limousine             1995.0           manuell   
7    1990.0  control    limousine             1998.0           manuell   
8     250.0     test          NaN             2000.0           manuell   
9     590.0  control          bus             1997.0           manuell   
11    350.0  control          NaN             2000.0               NaN   
12   5299.0  control   kleinwagen             2010.0         automatik   
13   1350.0  control    limousine             1999.0           manuell   
14   3999.0     test   kleinwagen             2007.0           manuell   
15  18900.0     test        coupe             1982.0         automatik   
16    350.0     test        kombi             1999.0           manuell   
17   5500.0     test        coupe             1990.0           manuell   
18    300.0  control          bus             1995.0           manuell   
19   4150.0  control          suv             2004.0           manuell   
20   3500.0     test        kombi             2003.0           manuell   
21  41500.0     test       cabrio             2004.0           manuell   
22  25450.0  control       cabrio             2015.0           manuell   
23   7999.0  control          bus             2010.0           manuell   
24  48500.0  control    limousine             2014.0         automatik   
25     90.0  control        kombi             1996.0           manuell   

    horse_power     model  odometer_km  registration_mont

Looking at some categorical data, some of it is in German, let's convert them to English counterparts.

In [37]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [38]:
mapping_dict = {'bus': 'bus',
                'limousine' : 'limo',
                'kleinwagen' : 'small car',
                'kombi' : 'station wagon',
                'nan' : 'nan',
                'coupe' : 'sedan',
                'suv' : 'suv'
}

autos['vehicle_type'] = autos['vehicle_type'].map(mapping_dict)

In [42]:
autos.head(25)

date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20  2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21  2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22  2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23  2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24  2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25  2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   

      price  ab_test   vehicle_type  registration_year transmission_type  \
0    5000.0  control            bus             2004.0           manuell   
1    8500.0  control           limo             1997.0         automatik   
2    8990.0     test           limo             2009.0           manuell   
3    4350.0  control      small car             2007.0         automatik   
4    1350.0     test  station wagon             2003.0           manuell   
5    7900.0     test            bus             2006.0         automatik   
6     300.0     test           limo             1995.0           manuell   
7    1990.0  control           limo             1998.0           manuell   
8     250.0     test            NaN             2000.0           manuell   
9     590.0  control            bus             1997.0           manuell   
11    350.0  control            NaN             2000.0               NaN   
12   5299.0  control      small car             2010.0         automatik   
13   1350.0  control           limo             1999.0           manuell   
14   3999.0     test      small car             2007.0           manuell   
15  18900.0     test          sedan             1982.0         automatik   
16    350.0     test  station wagon             1999.0           manuell   
17   5500.0     test          sedan             1990.0           manuell   
18    300.0  control            bus             1995.0           manuell   
19   4150.0  control            suv             2004.0           manuell   
20   3500.0     test  station wagon             2003.0           manuell   
21  41500.0     test            NaN             2004.0           manuell   
22  25450.0  control            NaN             2015.0           manuell   
23   7999.0  control            bus             2010.0           manuell   
24  48500.0  control           limo             2014.0         automatik   
25     90.0  control  station wagon             1996.0           manuell   

    h

In [43]:
transmission_dict = {'manuell' : 'manual',
                    'automatik' : 'automatic'}

autos['transmission_type'] = autos['transmission_type'].map(transmission_dict)

In [47]:
autos['transmission_type'].unique()

array(['manual', 'automatic', nan], dtype=object)

In [48]:
autos['unrepaired_damage'] = autos['unrepaired_damage'].map({'nein' : 'no', 'ja' : 'yes'})

Now that we replaced German Words with English ones for the vehicle and transmission type columns, let's look to convert the dates to be uniform numeric data

In [57]:
autos['date_crawled'] = autos['date_crawled'].str.replace('-', '')
autos['ad_created'] = autos['ad_created'].str.replace('-','')
autos['last_seen'] = autos['last_seen'].str.replace('-','')

Now let's look to find the most common brand/model combinations. 

In [75]:
brand_list = autos['brand'].value_counts().index
print(brand_list)
autos['brand'].describe()

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'jaguar', 'trabant', 'daewoo', 'rover', 'lancia',
       'lada'],
      dtype='object')


count          48016
unique            40
top       volkswagen
freq           10185
Name: brand, dtype: object

In [77]:
brand_model = {}
for b in brand_list:
    brands = autos[autos['brand']==b]
    desc_model = brands['name'].describe()
    brand_model[b] = desc_model[2]

for b in brand_model:
    print(b, ':', brand_model[b])

honda : Honda_Civic
chevrolet : Chevrolet_Spark_1.0_LS
smart : Smart_ForTwo
opel : Opel_Corsa
peugeot : Peugeot_206
jaguar : Jaguar_XF_2.2_Diesel
suzuki : Suzuki_Swift_1.6_Sport
fiat : Fiat_Punto
volvo : Volvo_V70_2.5
hyundai : Hyundai_i30_1.4_Classic
trabant : Trabant_601
mitsubishi : Mitsubishi_Colt
volkswagen : Volkswagen_Golf_1.4
porsche : Porsche_Boxster
land_rover : Land_Rover_Freelander_1.8i
rover : Rover_214_TÜV_8/2016_Auto
daewoo : Daewoo_Matiz
bmw : BMW_316i
citroen : Citroën_C1_1.0_Style
mini : MINI_Mini_One
mazda : Mazda_MX_5_1.6i_16V
toyota : Toyota_Aygo
dacia : Dacia_Logan_MCV_1.6_16V_Laureate
audi : Audi_A4_Avant_2.0_TDI_DPF
nissan : Nissan_Micra
jeep : Jeep_Grand_Cherokee_3.0_CRD_Automatik_Limited
chrysler : Chrysler_Stratus_2.5_LX
lada : Lada_Niva
renault : Renault_Twingo
alfa_romeo : Alfa_Romeo_147
subaru : Subaru_Legacy_2.5_4WD_Automatik_GX
kia : Kia_Picanto
saab : Saab_Cabrio_V6
skoda : Skoda_Fabia
sonstige_autos : Dodge_RAM
ford : Ford_Fiesta
seat : Seat_Ibiza
merc

Finally, let's determine if cars with damage are cheaper than those who are not

In [79]:
autos['unrepaired_damage'].unique()


array(['no', nan, 'yes'], dtype=object)

In [80]:
damaged_cars = autos[autos['unrepaired_damage']=='yes']
undamaged_cars = autos[autos['unrepaired_damage']=='no']
mean_price_damaged = damaged_cars['price'].mean()
mean_price_undamaged = undamaged_cars['price'].mean()

print('Mean Price of Damaged Cars is ', mean_price_damaged)
print('\n')
print('Mean Price of Undamaged Cars is ', mean_price_undamaged)

Mean Price of Damaged Cars is  2126.840091973244


Mean Price of Undamaged Cars is  7077.225495634909


We can indeed conclude that on average, damaged cars cost about 5000 less than that of undamaged.